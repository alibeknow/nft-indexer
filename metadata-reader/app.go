package metadata_reader

import (
	"context"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/zeebo/errs"
	"go.uber.org/zap"
	"golang.org/x/sync/errgroup"

	"metadata-reader/database"
	"metadata-reader/private/httpreader"
	"metadata-reader/queue"
	"metadata-reader/server"
)

var (
	requestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "request_per_second_gauge",
			Help: "Number of requests per second",
		},
	)
	messagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "message_per_second_gauge",
			Help: "Number of messages per second",
		},
	)
	instantRequestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_request_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)

	instantMessagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_messages_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)
)

func init() {
	prometheus.MustRegister(requestPerSecond)
	prometheus.MustRegister(messagePerSecond)
	prometheus.MustRegister(instantRequestPerSecond)
	prometheus.MustRegister(instantMessagePerSecond)
}

type App struct {
	log      *zap.Logger
	db       *database.DB
	srv      *server.Server
	listener *queue.Listener
}

func NewApp(log *zap.Logger, db *database.DB, srv *server.Server, listener *queue.Listener) (*App, error) {
	app := &App{
		log:      log,
		db:       db,
		srv:      srv,
		listener: listener,
	}

	return app, nil
}

func (mr *App) Run(ctx context.Context) error {
	group, ctx := errgroup.WithContext(ctx)

	group.Go(func() error {
		now := time.Now()
		ticker := time.NewTicker(time.Second)
		defer ticker.Stop()

		var (
			reqLast int64
			msgLast int64
		)

		for {
			select {
			case <-ctx.Done():
				return nil
			case <-ticker.C:
				duration := time.Since(now).Seconds()

				reqCount := httpreader.TotalRequestCount()
				requestPerSecond.Set(float64(reqCount) / duration)
				reqDelta := reqCount - reqLast
				reqLast = reqCount
				instantRequestPerSecond.Set(float64(reqDelta))

				msgCount := queue.TotalMessageCount()
				messagePerSecond.Set(float64(msgCount) / duration)
				msgDelta := msgCount - msgLast
				msgLast = msgCount
				instantMessagePerSecond.Set(float64(msgDelta))
			}
		}
	})
	group.Go(func() error {
		return mr.srv.Run(ctx)
	})
	group.Go(func() error {
		return mr.listener.Run(ctx)
	})

	return group.Wait()
}

func (mr *App) Close() error {
	serverErr := mr.srv.Close()
	listenerErr := mr.listener.Close()

	return errs.Combine(serverErr, listenerErr)
}
