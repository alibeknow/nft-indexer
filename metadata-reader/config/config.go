package config

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/spf13/viper"

	"metadata-reader/database"
	"metadata-reader/metadata"
	"metadata-reader/private/logutils"
	"metadata-reader/private/opensearch"
	"metadata-reader/queue"
	"metadata-reader/server"
)

type Config struct {
	Logger   logutils.Config
	Server   server.Config
	Database database.Config
	Queue    queue.Config
	Metadata metadata.Config
	Search   opensearch.Config
}

func LoadConfig(configPaths ...string) (Config, error) {
	v := viper.New()
	v.SetConfigName("config")
	v.SetConfigType("yaml")
	v.SetEnvPrefix("MR")
	v.SetEnvKeyReplacer(strings.NewReplacer(`.`, `_`))
	v.AutomaticEnv()

	v2 := viper.New()
	v2.SetConfigName("config")
	v2.SetConfigType("yaml")
	var buff bytes.Buffer
	m := map[string]interface{}{}
	err := json.NewEncoder(&buff).Encode(Config{})
	if err != nil {
		return Config{}, fmt.Errorf("failed to encode config: %v", err)
	}
	err = json.NewDecoder(&buff).Decode(&m)
	if err != nil {
		return Config{}, fmt.Errorf("failed to decode config: %v", err)
	}
	for key, val := range m {
		v2.Set(key, val)
	}

	for _, path := range configPaths {
		v.AddConfigPath(path)
		v2.AddConfigPath(path)
	}

	if err = v2.SafeWriteConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileAlreadyExistsError); ok {
			log.Println("config.yaml file already exists: skip creating")
		} else {
			return Config{}, fmt.Errorf("failed to write the configuration file: %v", err)
		}
	}

	if err = v.ReadInConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
			log.Println("config.yaml ignored")
		} else {
			return Config{}, fmt.Errorf("failed to read the configuration file: %v", err)
		}
	}

	for _, k := range v.AllKeys() {
		value := v.GetString(k)
		v.Set(k, os.ExpandEnv(value))
	}

	var config Config
	err = v.Unmarshal(&config)
	return config, err
}
