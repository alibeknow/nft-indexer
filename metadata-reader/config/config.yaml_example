server:
  address: ":xxx"

database:
  name: "xxx"
  protocol: "mongodb"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "mongodb" for docker
  port: xxx #number
  sslEnable: true/false
  directConnection: true/false

queue:
  protocol: "amqp"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "rabbitmq" for docker
  port: xxx #number
  name: "xxx"

reader:
  workers: xxx #number
  proxyUri: "xxx"
  ipfsGateway: "https://ipfs.io/ipfs/"
  pinataDefaultGateway: "https://gateway.pinata.cloud/ipfs/"

httpreader:
  retries: xxx #number
