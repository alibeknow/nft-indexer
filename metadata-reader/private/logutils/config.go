package logutils

type Config struct {
	Level string
}
