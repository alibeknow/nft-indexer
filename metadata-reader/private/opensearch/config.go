package opensearch

type Config struct {
	Uri      string
	Shards   int
	Replicas int
}
