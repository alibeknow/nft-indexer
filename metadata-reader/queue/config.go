package queue

import "fmt"

type Config struct {
	Protocol string
	Username string
	Password string
	Host     string
	Port     int
	Name     string
	Workers  int
}

func (config *Config) URL() string {
	url := fmt.Sprintf(
		"%s://%s:%s@%s:%d",
		config.Protocol,
		config.Username,
		config.Password,
		config.Host,
		config.Port,
	)

	return url
}
