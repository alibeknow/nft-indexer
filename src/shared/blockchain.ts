export enum Blockchain {
  ETH = 'eth',
  POLYGON = 'polygon',
}
