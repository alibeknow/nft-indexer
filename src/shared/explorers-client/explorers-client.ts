import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { IBaseConfig } from '@shared/baseconfig';
import { IExplorersClientConfig } from '@shared/explorers-client/explorers-client.config';
import axios, { AxiosResponse } from 'axios';
import { ethers } from 'ethers';
import { Blockchain } from '@shared/blockchain';

/**
 * ExplorerEvent
 */
export interface ExplorerEvent extends ethers.providers.Log {
  signature: string;
  contract: string;
}

/**
 * ExplorersClient
 */
@Injectable()
export class ExplorersClient<T extends IBaseConfig> {

  private readonly apiUri?: string;

  /**
   * ExplorersClient constructor
   * @param {ConfigService}
   */
  constructor(
    private configService: ConfigService<T>,
  ) {
    const explorersClientConfig = this.configService.get<IExplorersClientConfig>('explorersClient') as IExplorersClientConfig;
    this.apiUri = explorersClientConfig.explorersApi;
  }

  /**
   * Makes request to Explorers API endpoint
   * @param {string}
   */
  public async request<T>(uri: string): Promise<T[]> {
    let paginationToken: string;
    let results: T[] = [];

    const { data: result }: AxiosResponse = await axios.get<string>(
      `${this.apiUri}/${uri}`,
    );

    results = results.concat(...result.data);
    paginationToken = result.token;

    while(paginationToken !== null) {
      const { data: result }: AxiosResponse = await axios.get<string>(
        `${this.apiUri}/${uri}?token=${paginationToken}`,
      );

      results = results.concat(...result.data);
      paginationToken = result.token;
    }

    return results;
  }

  /**
   * Fetches wallet events by event signature
   * @param {Blockchain}
   * @param {string}
   * @param {string}
   */
  public async fetchWalletEvents(chainName: Blockchain, walletAddress: string, eventSignature: string): Promise<ExplorerEvent[]> {
    return this.request<ExplorerEvent>(`${chainName}/address/${walletAddress}/events/${eventSignature}`);
  }
}
