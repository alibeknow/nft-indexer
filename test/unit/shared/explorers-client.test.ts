import axios from 'axios';
import { ethers } from 'ethers';
import { ConfigService } from '@nestjs/config';
import { ExplorersClient, IExplorersClientConfig } from '@shared/explorers-client';
import { IApiConfig } from '@api/app.config';
import { Blockchain } from '@shared/blockchain';
import { explorersApiAxiosResponse1, explorersApiAxiosResponse2, explorersApiResponse } from './fixtures';

describe('ExplorersClient', () => {
  let explorersClient: ExplorersClient<IApiConfig>;
  let config: ConfigService<IApiConfig>;

  beforeAll(() => {
    config = {
      get: () => {
        return {
          explorersApi: 'http://explorers.api/blockchain/v4',
        } as IExplorersClientConfig;
      },
    } as unknown as ConfigService<IApiConfig>;

    explorersClient = new ExplorersClient(config);
  });

  describe('[fetchWalletEvents] method', () => {

    it('should return wallet events', async () => {
      const chainName = Blockchain.ETH;
      const walletAddress = '0x753F10598C026E73182CA74ED33DE05974b9f083';
      const eventSignature = ethers.utils.id('Transfer(address,address,uint256)');

      // @ts-ignore
      jest.spyOn(explorersClient, 'request').mockImplementationOnce(() => {
        return Promise.resolve(explorersApiResponse as unknown);
      });

      const result = await explorersClient.fetchWalletEvents(
        chainName, walletAddress, eventSignature,
      );

      expect(explorersClient.request).toHaveBeenNthCalledWith(1, `${chainName}/address/${walletAddress}/events/${eventSignature}`);
      expect(result).toEqual(explorersApiResponse);
    });
  });

  describe('[request] method', () => {
    it('should return list of results from explorers API', async () => {
      const chainName = Blockchain.ETH;
      const walletAddress = '0x753F10598C026E73182CA74ED33DE05974b9f083';
      const eventSignature = ethers.utils.id('Transfer(address,address,uint256)');

      jest.spyOn(axios, 'get').mockImplementation((url: string) => {
        return new Promise((resolve) => {
          setImmediate(() => {
            if(url === `${config.get('explorersClient').explorersApi}/${chainName}/address/${walletAddress}/events/${eventSignature}`) {
              return resolve(explorersApiAxiosResponse1);
            } else {
              return resolve(explorersApiAxiosResponse2);
            }
          });
        });
      });

      const results = await explorersClient.request(`${chainName}/address/${walletAddress}/events/${eventSignature}`);

      expect(results).toHaveLength(6);
      expect(axios.get).toHaveBeenCalledTimes(2);
      expect(results).toMatchSnapshot();
    });
  });
});
