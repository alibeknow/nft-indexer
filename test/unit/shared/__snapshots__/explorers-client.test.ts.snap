// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExplorersClient [request] method should return list of results from explorers API 1`] = `
[
  {
    "blockHash": "0x5465b913d75e5027b34321a1c738bc028696fa101076407050d45890fdd1d34f",
    "blockHeight": 15091714,
    "contract": "0x753f10598c026e73182ca74ed33de05974b9f083",
    "data": "0x",
    "logIndex": 127,
    "signature": "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
    "topics": [
      "0x0000000000000000000000000000000000000000000000000000000000000000",
      "0x00000000000000000000000047ba73787ea555cafbf1c4472b03471c73e46e76",
      "0x0000000000000000000000000000000000000000000000000000000000003149",
    ],
    "transactionHash": "0x0b70aeddc58087db0982e5a3299cf406f6cb67fe41df6b0bfc3ace4bbe8e4da3",
    "transactionIndex": 29,
  },
  {
    "blockHash": "0x5465b913d75e5027b34321a1c738bc028696fa101076407050d45890fdd1d34f",
    "blockHeight": 15091714,
    "contract": "0x753f10598c026e73182ca74ed33de05974b9f083",
    "data": "0x",
    "logIndex": 126,
    "signature": "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
    "topics": [
      "0x0000000000000000000000000000000000000000000000000000000000000000",
      "0x00000000000000000000000047ba73787ea555cafbf1c4472b03471c73e46e76",
      "0x0000000000000000000000000000000000000000000000000000000000003148",
    ],
    "transactionHash": "0x0b70aeddc58087db0982e5a3299cf406f6cb67fe41df6b0bfc3ace4bbe8e4da3",
    "transactionIndex": 29,
  },
  {
    "blockHash": "0x9a096f71e0b475675f87e150ce1f07080cb137d5a9d2b2f8bfbdc11725c86905",
    "blockHeight": 14269363,
    "contract": "0xd532b88607b1877fe20c181cba2550e3bbd6b31c",
    "data": "0x",
    "logIndex": 136,
    "signature": "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
    "topics": [
      "0x0000000000000000000000000c4a949315886f2cf07e1a54a277eed860f33711",
      "0x00000000000000000000000047ba73787ea555cafbf1c4472b03471c73e46e76",
      "0x0000000000000000000000000000000000000000000000000000000000001edc",
    ],
    "transactionHash": "0xeffc27ac18207071d542ab9e917ececd307f0f7c5b503c51b1712b1168172a27",
    "transactionIndex": 58,
  },
  {
    "blockHash": "0x02c6f9f733183a16662cbe449f84a3eed1850956c6afbeeee5f13d2bedd5d4ca",
    "blockHeight": 14237259,
    "contract": "0x4db1f25d3d98600140dfc18deb7515be5bd293af",
    "data": "0x",
    "logIndex": 26,
    "signature": "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
    "topics": [
      "0x000000000000000000000000f7f4e08f349a4f0610a44d18433dcdcc4a0e5b74",
      "0x00000000000000000000000047ba73787ea555cafbf1c4472b03471c73e46e76",
      "0x0000000000000000000000000000000000000000000000000000000000001d03",
    ],
    "transactionHash": "0x6e8859a2248b5a111217437db2a47ba175de854faf04ea8110e09296d78060c1",
    "transactionIndex": 34,
  },
  {
    "blockHash": "0xc8e4adf97e51cf2de5a1fd9fbf905284291241698bfe00a4d7224159b07a2cf4",
    "blockHeight": 14298517,
    "contract": "0x13d15d8b7b2bf48cbaf144c5c50e67b6b635b5cd",
    "data": "0x",
    "logIndex": 62,
    "signature": "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
    "topics": [
      "0x00000000000000000000000047ba73787ea555cafbf1c4472b03471c73e46e76",
      "0x0000000000000000000000002f554c24675623c194a4ec88ad39d062ab1779e4",
      "0x00000000000000000000000000000000000000000000000000000000000017e3",
    ],
    "transactionHash": "0xd3aee1d58291366bf861de810c3c3ef4b0f7a8a192727d55c58b140aa8033446",
    "transactionIndex": 47,
  },
  {
    "blockHash": "0xe2b13fba7e061af10e3c6f3f09bd464f769f72c9d3f4f044fc58cb4453096df3",
    "blockHeight": 14289024,
    "contract": "0x13d15d8b7b2bf48cbaf144c5c50e67b6b635b5cd",
    "data": "0x",
    "logIndex": 441,
    "signature": "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
    "topics": [
      "0x0000000000000000000000000000000000000000000000000000000000000000",
      "0x00000000000000000000000047ba73787ea555cafbf1c4472b03471c73e46e76",
      "0x00000000000000000000000000000000000000000000000000000000000017e5",
    ],
    "transactionHash": "0xe564e61431bb14b952de01c5fbdc05bd8b24778c801a46309a9fb3d32ed9ea9b",
    "transactionIndex": 355,
  },
]
`;
